daemon/k8s_watcher.go

```diff
@@ -771,356 +758,115 @@ func (d *Daemon) missingK8sNetworkPolicyV1(m versioned.Map) versioned.Map {
-		if _, ok := newSI.Ports[loadbalancer.FEPortName(port.Name)]; !ok {
-			newSI.Ports[loadbalancer.FEPortName(port.Name)] = p
-		}
-	}
-	return newSI
-}
+		switch event.Action {
+		case k8s.UpdateService, k8s.UpdateIngress:
+			if err := d.addK8sSVCs(event.ID, svc, event.Endpoints); err != nil {
+				scopedLog.WithError(err).Error("Unable to add/update service to implement k8s event")
+			}
 
-func (d *Daemon) addK8sServiceV1(svc *v1.Service) error {
-	newSI := parseSvcV1(svc)
-	if newSI == nil {
-		return nil
-	}
+			if svc.IsExternal() {
+				serviceImportMeta, cacheOK := endpointMetadataCache.get(event.ID)
+
+				// If this is the first time adding this Endpoint, or there was an error
+				// adding it last time, then try to add translate it and its
+				// corresponding external service for any toServices rules which
+				// select said service.
+				if !cacheOK || (cacheOK && serviceImportMeta.ruleTranslationError != nil) {
+					translator := k8s.NewK8sTranslator(event.ID, *event.Endpoints, false, svc.Labels, bpfIPCache.IPCache)
+					result, err := d.policy.TranslateRules(translator)
+					endpointMetadataCache.upsert(event.ID, err)
+					if err != nil {
+						log.Errorf("Unable to repopulate egress policies from ToService rules: %v", err)
+						break
+					} else if result.NumToServicesRules > 0 {
+						// Only trigger policy updates if ToServices rules are in effect
+						d.TriggerPolicyUpdates(true, "Kubernetes service endpoint added")
+					}
+				} else if serviceImportMeta.ruleTranslationError == nil {
+					d.TriggerPolicyUpdates(true, "Kubernetes service endpoint updated")
+				}
+			}
 
-	svcns := loadbalancer.K8sServiceNamespace{
-		ServiceName: svc.ObjectMeta.Name,
-		Namespace:   svc.ObjectMeta.Namespace,
-	}
+		case k8s.DeleteService, k8s.DeleteIngress:
+			if err := d.delK8sSVCs(event.ID, event.Service, event.Endpoints); err != nil {
+				scopedLog.WithError(err).Error("Unable to delete service to implement k8s event")
+			}
 
-	d.loadBalancer.K8sMU.Lock()
-	defer d.loadBalancer.K8sMU.Unlock()
+			if svc.IsExternal() {
+				endpointMetadataCache.delete(event.ID)
 
-	if oldSI, ok := d.loadBalancer.K8sServices[svcns]; ok {
-		if oldSI.Equals(newSI) {
-			return nil
+				translator := k8s.NewK8sTranslator(event.ID, *event.Endpoints, true, svc.Labels, bpfIPCache.IPCache)
+				result, err := d.policy.TranslateRules(translator)
+				if err != nil {
+					log.Errorf("Unable to depopulate egress policies from ToService rules: %v", err)
+					break
+				} else if result.NumToServicesRules > 0 {
+					// Only trigger policy updates if ToServices rules are in effect
+					d.TriggerPolicyUpdates(true, "Kubernetes service endpoint deleted")
+				}
+			}
 		}
 	}
+}
 
-	d.loadBalancer.K8sServices[svcns] = newSI
+func (d *Daemon) runK8sServiceHandler() {
+	go d.k8sServiceHandler()
+}
 
-	return d.syncLB(&svcns, nil, nil)
+func (d *Daemon) addK8sServiceV1(svc *v1.Service) error {
+	d.k8sSvcCache.UpdateService(svc)
```

Comment:
Is this a blocking operation?

Reply:
Generally no. It is very unlikely to block. If it blocks, we should stop processing more events and throttle down.